### PR TITLE
EdgeConnect status of CR doesnt show real situation, when pod is failing

### PR DESCRIPTION
**Pull request by andriisoldatenko**
## Description

THis PR is bugfix for https://dt-rnd.atlassian.net/browse/K8S-11360 

Before this PR:

```
kaf ../samples/edgeconnect_****_v1alpha1.yaml
edgeconnect.dynatrace.com/e2e-andrii created
secret/edge-connect-client-secret-***** unchanged
```

```
➜  $ k get ec
NAME         APISERVER                             STATUS    AGE
e2e-andrii   *****.dev.apps.dynatracelabs.com   Running   5s
```

but pod in crashloop state.

After this change status will be Deploying, because we are not reached desired amount of replicas.

## How can this be tested?

Deploy different edgeConnects, especially broken ones.

Make sure CR status represents the actual state of EdgeCOnnect Deployment.

### CODE REVIEW

File: pkg/controllers/edgeconnect/phase_test.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+		}
+
+		fakeClient := fake.NewClient(objects...)
+
+		controller := &Controller{
+			client:    fakeClient,
+			apiReader: fakeClient,
+		}
+		phase := controller.determineEdgeConnectPhase(ec)
+		assert.Equal(t, status.Deploying, phase)
```
Comment (aorcholski) [inline pkg/controllers/edgeconnect/phase_test.go:72]:
```suggestion
		assert.Equal(t, status.Running, phase)
```

File: pkg/controllers/edgeconnect/phase.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+	"github.com/Dynatrace/dynatrace-operator/pkg/api/status"
+	"github.com/Dynatrace/dynatrace-operator/pkg/api/v1alpha2/edgeconnect"
+	appsv1 "k8s.io/api/apps/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func (controller *Controller) determineEdgeConnectPhase(ec *edgeconnect.EdgeConnect) status.DeploymentPhase {
+	components := []func(ec *edgeconnect.EdgeConnect) status.DeploymentPhase{
```
Comment (aorcholski) [inline pkg/controllers/edgeconnect/phase.go:14]: It's not needed considering the current design of EC feature. No more components in the near future.

File: pkg/controllers/edgeconnect/phase_test.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+			createDeployment(testNamespace, "test-name", replicas, readyReplicas),
+		}
+
+		fakeClient := fake.NewClient(objects...)
+
+		controller := &Controller{
+			client:    fakeClient,
+			apiReader: fakeClient,
+		}
+		phase := controller.determineEdgeConnectPhase(ec)
```
Comment (aorcholski) [inline pkg/controllers/edgeconnect/phase_test.go:71]: ec.Spec.Replicas=0 so generally speaking  the test is broken

File: pkg/controllers/edgeconnect/phase_test.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+		assert.Equal(t, status.Error, phase)
+	})
+
+	t.Run("edgeConnect pods not ready -> deploying", func(t *testing.T) {
+		replicas, readyReplicas := int32(1), int32(0)
+		objects := []client.Object{
+			createDeployment(testNamespace, "test-name", replicas, readyReplicas),
```
Comment (aorcholski) [inline pkg/controllers/edgeconnect/phase_test.go:46]:
```suggestion
			createDeployment(testNamespace, testName, replicas, readyReplicas),
```

File: pkg/controllers/edgeconnect/phase_test.go (hunk 3 above)
Comment (andriisoldatenko) [reply pkg/controllers/edgeconnect/phase_test.go:71]: I don't get it, check line number 60-61. replicas == readyReplicas == 1

Comment (aorcholski) [reply pkg/controllers/edgeconnect/phase_test.go:71]: ec.Spec.Replicas is NIL so according to the `determineEdgeConnectPhase()`
```
scheduledReplicas := int32(0)
	if ec.Spec.Replicas != nil {
		scheduledReplicas = *deployment.Spec.Replicas
	}

	if scheduledReplicas != deployment.Status.ReadyReplicas {
```	
the fact that `deployment.replicas` = `deployment.readyReplicas` = 1 has nothing to do with the correct logic and final result

File: pkg/controllers/edgeconnect/phase_test.go (hunk 1 above)
Comment (andriisoldatenko) [reply pkg/controllers/edgeconnect/phase_test.go]: fixed in 93c4f1d0

File: pkg/controllers/edgeconnect/phase_test.go (hunk 3 above)
Comment (andriisoldatenko) [reply pkg/controllers/edgeconnect/phase_test.go:71]: fixed in 93c4f1d0

File: pkg/controllers/edgeconnect/phase_test.go (hunk 4 above)
Comment (andriisoldatenko) [reply pkg/controllers/edgeconnect/phase_test.go]: fixed in 93c4f1d0

File: pkg/controllers/edgeconnect/phase.go (hunk 2 above)
Comment (andriisoldatenko) [reply pkg/controllers/edgeconnect/phase.go]: fixed in 93c4f1d0

File: pkg/controllers/edgeconnect/phase.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+	err := controller.client.Get(context.Background(), types.NamespacedName{Name: ec.Name, Namespace: ec.Namespace}, deployment)
+	if k8serrors.IsNotFound(err) {
+		log.Info("edgeConnect deployment to be deployed", "edgeConnect", ec.Name, "deployment", ec.Name)
+
+		return status.Deploying
+	}
+
+	if err != nil {
+		log.Error(err, "edgeConnect deployment could not be accessed", "edgeConnect", ec.Name, "deployment", ec.Namespace)
```
Comment (aorcholski) [inline pkg/controllers/edgeconnect/phase.go:24]:
```suggestion
		log.Error(err, "edgeConnect deployment could not be accessed", "edgeConnect", ec.Name, "namespace", ec.Namespace)
```

File: pkg/controllers/edgeconnect/phase.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+		return status.Error
+	}
+
+	scheduledReplicas := int32(0)
+	if deployment.Spec.Replicas != nil {
+		scheduledReplicas = *deployment.Spec.Replicas
+	}
+
+	if scheduledReplicas != deployment.Status.ReadyReplicas {
+		log.Info("edgeConnect deployment is still deploying", "edgeConnect", ec.Name, "deployment", ec.Name)
```
Comment (aorcholski) [inline pkg/controllers/edgeconnect/phase.go:35]:
```suggestion
		log.Info("edgeConnect deployment is still deploying", "edgeConnect", ec.Name, "namespace", ec.Name)
```

File: pkg/controllers/edgeconnect/phase.go (hunk 5 above)
Comment (andriisoldatenko) [reply pkg/controllers/edgeconnect/phase.go]: fixed in eaeb5dd7

File: pkg/controllers/edgeconnect/phase.go (hunk 6 above)
Comment (andriisoldatenko) [reply pkg/controllers/edgeconnect/phase.go]: fixed in eaeb5dd7

**Review by aorcholski — APPROVED** [review]
LGTM, failing EC POD -> EC.Status.`Deploying`

**Review by luhi-DT — APPROVED** [review]
LGTM

**Pull request merged**